src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,443 @@
+     * The composer's status will be at least BUILT after executing this method.
+     */
+    public List<TransactionWithSigner> buildGroup() throws IOException {
+        if (this.status.compareTo(Status.BUILT) >= 0)
+            return this.transactionList;
+
+        if (this.transactionList.size() == 0)
+            throw new IllegalArgumentException("should not build transaction group with 0 transaction in composer");
+
+        List<Transaction> groupTxns = new ArrayList<>();
+        for (TransactionWithSigner t : this.transactionList) groupTxns.add(t.txn);
+        Digest groupID = TxGroup.computeGroupID(groupTxns.toArray(new Transaction[0]));
+        for (TransactionWithSigner tws : this.transactionList) tws.txn.group = groupID;
+
+        this.status = Status.BUILT;
+        return this.transactionList;
+    }
+
+    /**
+     * Obtain signatures for each transaction in this group. If signatures have already been obtained,
+     * this method will return cached versions of the signatures.
+     * <p>
+     * The composer's status will be at least SIGNED after executing this method.
+     * <p>
+     * An error will be thrown if signing any of the transactions fails.
+     *
+     * @return an array of signed transactions.
+     */
+    public List<SignedTransaction> gatherSignatures() throws IOException, NoSuchAlgorithmException {
+        if (this.status.compareTo(Status.SIGNED) >= 0)
+            return this.signedTxns;
+
+        List<TransactionWithSigner> txnAndSignerList = this.buildGroup();
+
+        HashMap<TxnSigner, List<Integer>> transSignerToIndices = new HashMap<>();
+        List<SignedTransaction> tempSignedTxns = new ArrayList<>();
+        for (int i = 0; i < txnAndSignerList.size(); i++) tempSignedTxns.add(null);
+
+        for (int i = 0; i < txnAndSignerList.size(); i++) {
+            TransactionWithSigner tws = txnAndSignerList.get(i);
+            if (transSignerToIndices.containsKey(tws.signer)) {
+                List<Integer> curr = transSignerToIndices.get(tws.signer);
+                curr.add(i);
+                transSignerToIndices.put(tws.signer, curr);
+            } else {
+                List<Integer> init = new ArrayList<>();
+                init.add(i);
+                transSignerToIndices.put(tws.signer, init);
+            }
+        }
+
+        Transaction[] txnGroup = new Transaction[txnAndSignerList.size()];
+        for (int i = 0; i < txnAndSignerList.size(); i++)
+            txnGroup[i] = txnAndSignerList.get(i).txn;
+
+        for (TxnSigner txnSigner : transSignerToIndices.keySet()) {
+            List<Integer> indices = transSignerToIndices.get(txnSigner);
+            int[] indicesPrimitive = ArrayUtils.toPrimitive(indices.toArray(new Integer[0]));
+            SignedTransaction[] signed = txnSigner.signTxnGroup(txnGroup, indicesPrimitive);
+            for (int i = 0; i < indicesPrimitive.length; i++)
+                tempSignedTxns.set(indicesPrimitive[i], signed[i]);
```

Comment:
nit: we might want to check if `tempSignedTxns` has a `null` value in case some signer misbehaves